### PR TITLE
Separate outputs for better quantization performance

### PR DESCRIPTION
**Pull request by shashichilappagari**
### Summary
🚀🛠️🐛

This pull request adds features and improvements to export YOLOv8 models to TensorFlow SavedModels and TensorFlow Lite models, with options to optimize the models for hardware devices and quantize them to integer data types. It modifies several files in the `ultralytics` package, such as `exporter.py`, `predict.py`, `val.py`, `block.py`, `head.py`, and `autobackend.py`, and adds new files `quantize.py` and `replace.json` to support the export and quantization process. It also adds a new model configuration file `relu6-yolov8.yaml` for YOLOv8 with ReLU6 activation.

> _We're sailing on the edge with our YOLOv8_
> _We need to export it right, so don't you hesitate_
> _We'll tweak the `C2f` and the `Detect` head_
> _And quantize the model with `uint8` instead_

### Walkthrough
*  Add new arguments `separate_6_outputs` and `export_hw_optimized` to the default configuration file to control the export behavior of the YOLOv8 model ([link](https://github.com/ultralytics/ultralytics/pull/4495/files?diff=unified&w=0#diff-732d61cb4969eff986cec58058000b2949ea3b429a97ca5a1045897e754145c5L81-R82))
*  Add new model configuration file for YOLOv8 with ReLU6 activation, which can improve the quantization performance of the model ([link](https://github.com/ultralytics/ultralytics/pull/4495/files?diff=unified&w=0#diff-2f59e45f92d8b22a693f055610b8967e61672d238337b9f9e8f26cf77ab23216R1-R48))
*  Modify the `Exporter` class to pass the `separate_6_outputs` argument to the `C2f` module, which determines whether to split or concatenate its output ([link](https://github.com/ultralytics/ultralytics/pull/4495/files?diff=unified&w=0#diff-b9f4cf4260fabd8ffb31e9a85b332a63d39fb49829ce1d98d980b2ec3549322cL197-R204))
*  Modify the `Exporter` class to pass a `replace_json` file to the `onnx2tf` command, which contains a list of operations and parameters that need to be replaced or modified in the ONNX model ([link](https://github.com/ultralytics/ultralytics/pull/4495/files?diff=unified&w=0#diff-b9f4cf4260fabd8ffb31e9a85b332a63d39fb49829ce1d98d980b2ec3549322cL659-R665), [link](https://github.com/ultralytics/ultralytics/pull/4495/files?diff=unified&w=0#diff-12fc3a83e7cbc3dd5722b0386b0e7f0a5a119e27d072188df6d307aa8c30edcdL1-R28))
*  Add a new script `quantize.py` that performs full integer quantization on a TensorFlow SavedModel using a representative dataset ([link](https://github.com/ultralytics/ultralytics/pull/4495/files?diff=unified&w=0#diff-8562b61850a9c66fb53c4be93f90cf1cc25169fbe6c9cc42910dc0aa12ec401fL1-R39))
*  Modify the `Detect` class to add a boolean flag that indicates whether to export the model with 6 separate outputs or one concatenated output, and modify the `forward` method to handle both cases ([link](https://github.com/ultralytics/ultralytics/pull/4495/files?diff=unified&w=0#diff-06038a7c9e9c565760881118b974c2f2fe7aee7c78bb6c142d1498d0d9614639R29), [link](https://github.com/ultralytics/ultralytics/pull/4495/files?diff=unified&w=0#diff-06038a7c9e9c565760881118b974c2f2fe7aee7c78bb6c142d1498d0d9614639L46-R59), [link](https://github.com/ultralytics/ultralytics/pull/4495/files?diff=unified&w=0#diff-06038a7c9e9c565760881118b974c2f2fe7aee7c78bb6c142d1498d0d9614639L59-R71))
*  Add two new attributes to the `C2f` class that split the input tensor into two separate tensors, and add a new method `forward_hw_optimized` that optimizes the `C2f` module for hardware acceleration by avoiding concatenation and reducing the number of channels ([link](https://github.com/ultralytics/ultralytics/pull/4495/files?diff=unified&w=0#diff-70b1aad8c0068184a0e145431a251d90d480aed0226918c90c5a8ccd5a84f0f8R176-R177), [link](https://github.com/ultralytics/ultralytics/pull/4495/files?diff=unified&w=0#diff-70b1aad8c0068184a0e145431a251d90d480aed0226918c90c5a8ccd5a84f0f8L189-R212))
*  Add two import statements to the `predict.py` and `val.py` files, which import the `DFL` and `dist2bbox` modules that are used to decode the bounding box predictions from the model output ([link](https://github.com/ultralytics/ultralytics/pull/4495/files?diff=unified&w=0#diff-9dded00f5e51a32b00c6266aa445990bae22fe6ea0bbb9d2c8898f4fcdb8cb23R8-R9), [link](https://github.com/ultralytics/ultralytics/pull/4495/files?diff=unified&w=0#diff-023bb20ae0db3d69310690d02fbea0e2a6d57e5cf3878f00ea1cd097761fd7b5R8-R9))
*  Modify the `DetectionPredictor` and `DetectionValidator` classes to add a conditional branch that checks if the model output has 6 separate tensors, and if so, calls the `decode_bbox` method, which applies the `DFL` and `dist2bbox` functions to each pair of tensors and concatenates them into one tensor ([link](https://github.com/ultralytics/ultralytics/pull/4495/files?diff=unified&w=0#diff-9dded00f5e51a32b00c6266aa445990bae22fe6ea0bbb9d2c8898f4fcdb8cb23L24-R47), [link](https://github.com/ultralytics/ultralytics/pull/4495/files?diff=unified&w=0#diff-023bb20ae0db3d69310690d02fbea0e2a6d57e5cf3878f00ea1cd097761fd7b5R45), [link](https://github.com/ultralytics/ultralytics/pull/4495/files?diff=unified&w=0#diff-023bb20ae0db3d69310690d02fbea0e2a6d57e5cf3878f00ea1cd097761fd7b5L79-R99))
*  Modify the `AutoBackend` class to add `np.uint8` to the list of integer data types that indicate a quantized model, since some models may use unsigned 8-bit integers instead of signed 8-bit integers ([link](https://github.com/ultralytics/ultralytics/pull/4495/files?diff=unified&w=0#diff-2e3ad5da4ba57eb68a9b5aaddf0c40c784d888b7619ed337ce31dfea4f6380f9L404-R404))

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
This PR adds new export options for YOLO models, enabling "separate_outputs" and "export_hw_optimized" modes for improved quantization and hardware performance, along with related documentation and benchmarking updates. 🛠️🚀

---

### 📊 Key Changes
- **New Export Options:**  
  - `separate_outputs`: Exports model outputs separately to improve quantization and compatibility with certain deployment scenarios.
  - `export_hw_optimized`: Optimizes model blocks for faster inference on specific hardware.
- **Benchmark & CI Enhancements:**  
  - Benchmarks now test models with these new export options.
  - CI workflow updated to include these benchmarks.
- **Documentation Updates:**  
  - Export, predict, and validation docs now describe the new options and their usage.
  - Export format tables updated to show which formats support the new options.
- **Codebase Improvements:**  
  - Model export, prediction, and validation logic updated to handle new output modes.
  - Added utility functions for post-processing separate outputs.
  - Added new model YAML (ReLU6 variant) and JSON files for post-processing.
- **Compatibility Checks:**  
  - Ensures new options are only used with supported export formats.
- **Minor Fixes:**  
  - Improved handling of quantized TFLite models.
  - Small refactors and bug fixes in model and export logic.

---

### 🎯 Purpose & Impact
- **Better Quantization Support:**  
  - "Separate outputs" mode helps produce models that quantize more effectively, leading to smaller, faster, and more accurate models on edge devices.
- **Hardware Optimization:**  
  - "HW optimized" export mode tailors models for improved inference speed on certain hardware accelerators.
- **Expanded Deployment Flexibility:**  
  - Users can now export models in ways that better fit their deployment targets, especially for mobile and embedded use cases.
- **Clearer Documentation:**  
  - Users benefit from updated guides and tables, making it easier to understand export options and their compatibility.
- **Improved Testing:**  
  - Enhanced CI and benchmarking ensure new features are robust and reliable.

---

**In short:**  
This update gives users more control and flexibility when exporting YOLO models, especially for quantized and hardware-accelerated deployments, while also improving documentation and testing. 🚀💡

### CODE REVIEW

**Comment by shashichilappagari** [comment]
I have read the CLA Document and I sign the CLA

**Comment by franklin-degirum** [comment]
I have read the CLA Document and I sign the CLA

**Comment by shashichilappagari** [comment]
@glenn-jocher A quick update on the progress we made since last week. We have now generalized our approach to be applicable to segmentation and pose models as well. Classification models do not suffer from the quantization issues encountered by detection, segmentation, and pose models. Moreover, we have verified that the methods work for p6 models as well. We added tests to cover all the new code we developed and currently all tests are passing. We will write a guide in the docs to explain how to use the new options and also add the new options to the docs in appropriate places.

**Comment by shashichilappagari** [comment]
@glenn-jocher All tests are passing now. We also updated the documentation to include the new options. Please let us know if we need to do anything else for this pull request.

**Comment by glenn-jocher** [comment]
@shashichilappagari thank you for the update. I'm glad to hear all the tests are passing and that you've updated the documentation to include the new options. At this stage, your pull request will be reviewed by the maintainers. They may have further comments or suggestions. Once everything is satisfactory to them, your pull request will be merged. Thanks for your contribution to YOLOv8.

**Comment by shashichilappagari** [comment]
The modifications in this pull request improve the mAP of INT8 TFLite models substantially. Here are model evaluation results on the COCO val dataset (5000 images)

### Detection Models:
| Model              | Format          | separate_outputs          | mAP (val) 50:95     | mAP (val) 50  |
|-----------------|---------------|---------------------------|---------------------|--------------|
| yolov8n           | TFLite INT8    | False                              | 28.1                        | 45.3            |
| yolov8n           | TFLite INT8    | True                               | 34.6                        | 49.3             |
| yolov5n6u       | TFLite INT8    | False                              | 26.7                        | 43.4             |
| yolov5n6u       | TFLite INT8    | True                               |34.3                         |48.4              |

### Segmentation Models:
| Model              | Format          | separate_outputs          | mAP (box) 50:95    | mAP (mask) 50:95  |
|-----------------|---------------|---------------------------|---------------------|-----------------------|
| yolov8n-seg    | TFLite INT8   | False                              | 21.5                       | 22.3                          |
| yolov8n-seg    | TFLite INT8   | True                               | 34.3                       | 28.8                          |

### Pose Models:
| Model              | Format          | separate_outputs          | mAP (pose) 50:95    | mAP (pose) 50         |
|-----------------|---------------|---------------------------|---------------------  |-----------------------|
| yolov8n-pose  | TFLite INT8   | False                              | 0                              | 0                               |
| yolov8n-pose  | TFLite INT8   | True                               | 44.9                         | 77.7                          |

**Comment by glenn-jocher** [comment]
@shashichilappagari thank you for sharing the results of your pull request! It definitely looks promising. The enhancement in the mAP of INT8 TFLite models is substantial and beneficial to YOLOv8. Particularly, the fact that these improvements span across different model types — detection, segmentation, and pose models — shows how impactful your changes could be. We appreciate the effort you've put into testing these modifications and providing a comprehensive report. Now, we'll review the code changes and offer any further recommendations if required.

**Comment by shashichilappagari** [comment]
@glenn-jocher we understand that this is a PR with significant changes, and it takes time to review. We are syncing with the main every day to stay up to date. Please let us know if we can do anything to improve the PR.

**Comment by glenn-jocher** [comment]
@shashichilappagari, thank you for the effort that you've put into this pull request and for staying updated with the main repository. We appreciate your patience during this process as the review of such large changes can indeed take time. When reviewing your PR, we'll be looking to ensure clarity, continued ease of use, and maintainability of the codebase. We also appreciate tests performed to prove the benefits of the changes.

At this moment, simply continue to keep your PR synced with the main branch regularly. This helps us to incorporate your changes smoothly when the time comes. If there are any specific changes or clarifications needed on your PR, we will be sure to communicate them in the review.

Finally, thank you again for your significant contribution to improve YOLOv8. Your patience and engagement with the process is greatly appreciated.

**Comment by Skillnoob** [comment]
Would love for this to get merged as the posted testing results look very promising for a improved performance. Could be especially useful on devices like Raspberry Pi's or Phones which don't have a lot of compute.

**Comment by glenn-jocher** [comment]
Thank you for your comment. We also share your enthusiasm about the promising results shared by @Skillnoob. This enhancement could indeed improve the performance of YOLOv8 models substantially, making them more efficient on devices with limited computational resources like Raspberry Pis or smartphones.

Please note that these changes are currently under review. We're ensuring that the changes are robust, maintainable, and do not introduce unintended issues. We appreciate your patience and understanding in this process. As soon as the review is complete and everything is found satisfactory, we will merge this pull request.

File: ultralytics/utils/replace.json [hunk 1]
```diff
@@ -0,0 +1,109 @@
+{
```
Comment (Zoxive) [inline ultralytics/utils/replace.json:1]: Is there an extra step that needs to take place to include .json files in the pip package?

When i try installing this branch to test it, i noticed all the .py files are updated but `replace.json` and `post_replace.json` are missing.

I'm still new to the python ecosystem, but If i had to guess the `MANIFEST.in` needs to be updated.

Comment (shashichilappagari) [reply ultralytics/utils/replace.json:1]: @Zoxive I am not sure if we are supposed to answer this question. If so, can you explain how you are installing this branch for testing and what tests you are running?

Comment (Zoxive) [reply ultralytics/utils/replace.json:1]: I installed by `pip install git+https://github.com/DeGirum/ultralytics_yolov8.git@franklin_current`

I was able to successfully train a custom model with the new `relu6-yolov8.yaml`
When i attempted to export it with `format=tflite` it would complain that the `replace.json` was missing.

When then i double checked that i did actually have all the changes, and i manually checked out your banch into another folder and did a folder comparision with where pip installed the branch. I noticed that only the *`.json` files were missing.

Comment (shashichilappagari) [reply ultralytics/utils/replace.json:1]: @Zoxive Thanks for the details. We will check to see if adding the files to manifest helps or not. May I ask if you are the assigned reviewer for the pull request?

Comment (Zoxive) [reply ultralytics/utils/replace.json:1]: Nope. I'm just a bystander. I was just recently trying to export a tflite model and had problems, so i thought i would try out your branch especially to see if i could see the quantization improvements.

**Comment by Pbatch** [comment]
Is there any update on the review of these changes?

**Comment by glenn-jocher** [comment]
@Pbatch the review is ongoing, and we're working diligently on it. We'll share updates once the process is complete. Thanks for your patience.

**Comment by jbdebelle** [comment]
Hello @shashichilappagari,

Could you please provide some insight into the following features you've implemented:

The new separate_outputs export flag: How does this feature enhance quantization performance, and what specific optimizations does it enable for devices supporting hardware-accelerated inference?

The export_hw_optimized export flag: What optimizations are applied to the C2f block for faster inference on compatible hardware, and what kind of speed improvements can we expect to see?

Additionally, if you could recommend any resources or documentation to help me better understand the quantization process, that would be greatly appreciated.

Thank you in advance for your assistance!

Best regards,
